packages/openapi3/src/decorators.ts

```diff
@@ -0,0 +1,34 @@
+import {
+  DecoratorContext,
+  Program,
+  setDecoratorNamespace,
+  Type,
+  validateDecoratorTarget,
+} from "@cadl-lang/compiler";
+
+const refTargetsKey = Symbol();
+export function $useRef({ program }: DecoratorContext, entity: Type, refUrl: string): void {
```

Comment:
The `ref` decorators are used by `cadl-autorest`, I believe they will need to go to the shared `openapi` library instead

Reply:
The only thing I can think of is to have some sort of abstraction over these paths. Maybe something like placeholders for common locations with different paths. like `$schemas` expands to /components/schemas for v3 and /definitions for V2, etc. for other paths.

So instead of  `@useRef("../common.json#/definitions/Foo")` you would write  `@useRef("../common.json#$schemas/Foo")

I'm not sure this is better than the namespace though. It might be adding complexity for an uncommon case. I guess the problem is that useRef is not abstract enough, it refers to physical locations in OpenAPI. But I guess that's the point of it?

If useRef is not common, and I guess it isn't judging by our samples, then I think the namespacing solution is fine. But if it is common, then the experience won't be great.

All this to say +1 to the change as proposed from me, but maybe this gives others different ideas.

Reply:
I think it should really not be that common to do that. We currently only use that for the ARM common types but I think for a normal user it should probably not matter much.

Reply:
Yeah, I had the same comment when this wasn't moved into openapi, but have come around to the need for this. So, it is used by our common library definitions for azure-resource-manager in openapi2, but I don't see us having common swagger references in the Azure library for data plane.  I think when we support azure-inflected openapi3 for management plane, we might consider having an abstraction, but the abstraction may be that we use refs in openapi2 and just generate the library types as schema in openapi3.  Meanwhile, this allows us to use openapi decorators together.  Other uses of refs should be rare.